### PR TITLE
Refactored CMake host/target SDK detection to support cross-compiling

### PR DESCRIPTION
**Pull request by karwa**
#### What's in this pull request?
- `SWIFT_HOST_VARIANT_SDK` able to be given as build flag, existing calculation from `CMAKE_SYSTEM_NAME` refactored
- `SWIFT_HOST_VARIANT_ARCH` able to be given as build flag, existing calculation from `CMAKE_SYSTEM_PROCESSOR` refactored
- Calculation of host triple for Linux targets refactored, able to be given as build flag.
- Do not attempt to execute C source when cross-compiling. CMake won't let you, anyway.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -386,29 +386,74 @@ function(check_working_std_regex result_var_name)
...
+        set("${result_var_name}" TRUE PARENT_SCOPE)
+      else()
+        set("${result_var_name}" FALSE PARENT_SCOPE)
+      endif()
     endif()
   endif()
 endfunction()
 check_working_std_regex(SWIFT_HAVE_WORKING_STD_REGEX)
 
+# If SWIFT_HOST_VARIANT_SDK not given, try to detect from the CMAKE_SYSTEM_NAME
```
Comment (gribozavr) [inline CMakeLists.txt]: Please end each comment with a full stop.

File: CMakeLists.txt [hunk 2]
```diff
@@ -386,29 +386,74 @@ function(check_working_std_regex result_var_name)
...
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv6l")
+    set(SWIFT_HOST_VARIANT_ARCH "armv6")
+  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7l")
+    set(SWIFT_HOST_VARIANT_ARCH "armv7")
+  else()
+    message(FATAL_ERROR "Unrecognized architecture on host system: ${CMAKE_SYSTEM_PROCESSOR}")
+  endif()
+endif()
+
+set(SWIFT_HOST_VARIANT_SDK "${SWIFT_HOST_VARIANT_SDK}" CACHE STRING
```
Comment (gribozavr) [inline CMakeLists.txt]: Reusing the same variable name looks very confusing.  Could you change the detection logic above to be unconditional, and store the result into a different variable, say, `SWIFT_HOST_VARIANT_SDK_default`?

File: CMakeLists.txt [hunk 3]
```diff
@@ -435,114 +480,84 @@ function(is_sdk_requested name result_var_name)
   endif()
 endfunction()
 
-# FIXME: separate the notions of SDKs used for compiler tools and target
-# binaries.
-if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
-  set(CMAKE_EXECUTABLE_FORMAT "ELF")
+# Required from this point:
+# - SWIFT_HOST_VARIANT_SDK  - OS of system which will host the compiler
+# - SWIFT_HOST_VARIANT_ARCH - arch of system which will host the compiler
```
Comment (gribozavr) [inline CMakeLists.txt]: Sorry, I don't understand the point of the first part of this comment.  Doesn't it repeat what was already explained in the code above?  (In the `set(...CACHE...)` part.)

File: CMakeLists.txt [hunk 4]
```diff
@@ -435,114 +480,84 @@ function(is_sdk_requested name result_var_name)
...
+# 
+# FIXME: cross-compile standard libarary targets are specified by "SDK",
+# which locks each one to the hardcoded architectures defined here (e.g. armv7 for ANDROID)
```
Comment (gribozavr) [inline CMakeLists.txt]: Sorry, it is not clear what this FIXME means, what part of the behavior is undesirable, and how it should be changed.  Could you clarify?

File: CMakeLists.txt [hunk 5]
```diff
@@ -435,114 +480,84 @@ function(is_sdk_requested name result_var_name)
...
-    set(SWIFT_PRIMARY_VARIANT_ARCH_default "x86_64")
-
-    if("${SWIFT_ANDROID_NDK_PATH}" STREQUAL "")
-      set(swift_can_crosscompile_stdlib FALSE)
+  # Calculate the host triple
+  if("${SWIFT_HOST_TRIPLE}" STREQUAL "")
+    if("${SWIFT_HOST_VARIANT_ARCH}" MATCHES "(x86_64|aarch64|powerpc64|powerpc64le)")
+      set(SWIFT_HOST_TRIPLE "${SWIFT_HOST_VARIANT_ARCH}-unknown-${SWIFT_HOST_VARIANT}-gnu")
+    elseif("${SWIFT_HOST_VARIANT_ARCH}" MATCHES "(armv6|armv7)")
+      set(SWIFT_HOST_TRIPLE "${SWIFT_HOST_VARIANT_ARCH}-unknown-${SWIFT_HOST_VARIANT}-gnueabihf")
```
Comment (gribozavr) [inline CMakeLists.txt]: I would strongly prefer an explicit list of triples instead, like in the old code.  The reason is that once we need a single exception to this logic, or once a maintainer for one architecture needs to make a change, every other platform can break and it is not always clear how to modify the code without access to all the hardware.

File: CMakeLists.txt [hunk 6]
```diff
@@ -553,16 +568,19 @@ elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
   message(STATUS "${xcode_version}")
   message(STATUS "")
 
+  # Should we build the standard library for the host?
   is_sdk_requested(OSX swift_build_osx)
   if(swift_build_osx)
     configure_sdk_darwin(
         OSX "OS X" "${SWIFT_DARWIN_DEPLOYMENT_VERSION_OSX}"
-        macosx macosx macosx "x86_64")
+        macosx macosx macosx "${SWIFT_HOST_VARIANT_ARCH}")
```
Comment (gribozavr) [inline CMakeLists.txt]: Swift on OS X only supports x86_64, please revert this change for clarity.

File: CMakeLists.txt (hunk 4 above)
Comment (karwa) [inline CMakeLists.txt]: I mean that the parameters we specify in `SWIFT_SDKS` are lacking architecture specifics, so we need to hard-code it. For example, the SDK for Android is just `ANDROID`, which we assume below to be armv7. The iOS SDKs all have their architectures hardcoded because they are just specified by name (e.g. `IOS` or `WATCHOS`). We can't cross-compile the standard library for another linux architecture, because the SDK list would just be `LINUX` and we couldn't disambiguate it from the host.

To fix it, we would need to append the architecture to the SDK, for example: `OSX-x86_64;IOS-armv7`. We could easily do that - we have all of that information in `build-script-impl`.

Comment (gribozavr) [inline CMakeLists.txt]: I see.  This is something that we actually want even for iOS, to be able to enable/disable either SDK as a whole, or just particular slices.  Could you paste your comment into the patch as the FIXME text?

File: CMakeLists.txt [hunk 7]
```diff
@@ -435,114 +480,84 @@ function(is_sdk_requested name result_var_name)
...
       "Deployment OS for Swift host tools (the compiler) [windows].")
+  configure_sdk_unix(CYGWIN "Cygwin" "windows" "cygwin" "${SWIFT_HOST_VARIANT_ARCH}" "${SWIFT_HOST_VARIANT_ARCH}-unknown-${SWIFT_HOST_VARIANT}-cygnus")
 
-  set(SWIFT_HOST_VARIANT_SDK "CYGWIN")
-  set(SWIFT_HOST_VARIANT_ARCH "x86_64")
-  
-  set(SWIFT_PRIMARY_VARIANT_SDK_default "CYGWIN")
-  set(SWIFT_PRIMARY_VARIANT_ARCH_default "x86_64")
-elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
+elseif("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "OSX")
```
Comment (gribozavr) [inline CMakeLists.txt]: When cross-compiling the Swift compiler for watchOS, we used to enter this code path.  Would that still work after your change?

File: CMakeLists.txt [hunk 8]
```diff
@@ -435,114 +480,84 @@ function(is_sdk_requested name result_var_name)
...
-  configure_sdk_unix(CYGWIN "Cygwin" "windows" "cygwin" "x86_64" "x86_64-unknown-windows-cygnus")
-
-#  set(CMAKE_EXECUTABLE_FORMAT "ELF")
-
+  # FIXME: Using the host OS version won't produce correct results for
+  # cross-compilation.
+  string(REPLACE "[-].*" "" FREEBSD_SYSTEM_VERSION ${CMAKE_SYSTEM_VERSION})
+  message(STATUS "FreeBSD Version: ${FREEBSD_SYSTEM_VERSION}")
+  configure_sdk_unix(FREEBSD "FreeBSD" "freebsd" "${SWIFT_HOST_VARIANT}" "${SWIFT_HOST_VARIANT_ARCH}"
+    "${SWIFT_HOST_VARIANT_ARCH}-unknown-${SWIFT_HOST_VARIANT}${FREEBSD_SYSTEM_VERSION}")
```
Comment (gribozavr) [inline CMakeLists.txt]: I'd recommend changing back to hardcoding x86_64 and an explicit triple for now.  It is likely this code won't immediately work for cross-compilation on FreeBSD, and the person who would add this support, would likely find it easier to generalize concrete code than to re-abstact existing code that tries to be generic.

**Comment by gribozavr** [comment]
@swift-ci Please test

File: CMakeLists.txt (hunk 7 above)
Comment (karwa) [inline CMakeLists.txt]: It should do - this is all just about the host, and the iOS targets have always explicitly specified all of these settings in `build-script-impl`

Comment (gribozavr) [inline CMakeLists.txt]: Even the `configure_sdk_darwin` call that we'd be skipping?..

Comment (karwa) [inline CMakeLists.txt]: Why would we skip a call to `configure_sdk_darwin`? Do you mean a watchOS host? If so, and it was using the `${CMAKE_SYSTEM_NAME} STREQUAL "Darwin"` check from before, it should still get detected as `OSX`

Comment (gribozavr) [inline CMakeLists.txt]: The case I am thinking of is when we are on OS X, building a Swift compiler that would run on a watchOS device.  I think in that case `SWIFT_HOST_VARIANT_SDK` would be equal to `WATCHOS`, as set by `build-script-impl`.

Comment (karwa) [inline CMakeLists.txt]: Ah, yes you are correct - so I suppose all IOS SDK hosts should go down that path.

Comment (gribozavr) [inline CMakeLists.txt]: Yes, I think all iOS, tvOS, watchOS, native and simulators, should enter this path.

**Comment by gribozavr** [comment]
@swift-ci Please test OS X platform

File: CMakeLists.txt [hunk 9]
```diff
@@ -435,114 +484,91 @@ function(is_sdk_requested name result_var_name)
...
+  set(CMAKE_EXECUTABLE_FORMAT "ELF")
+  set(SWIFT_HOST_VARIANT "freebsd" CACHE STRING
+      "Deployment OS for Swift host tools (the compiler) [freebsd].")
+
   # FIXME: Using the host OS version won't produce correct results for
   # cross-compilation.
   string(REPLACE "[-].*" "" FREEBSD_SYSTEM_VERSION ${CMAKE_SYSTEM_VERSION})
   message(STATUS "FreeBSD Version: ${FREEBSD_SYSTEM_VERSION}")
-  configure_sdk_unix(FREEBSD "FreeBSD" "freebsd" "freebsd" "x86_64"
+  configure_sdk_unix(FREEBSD "FreeBSD" "freebsd" "${SWIFT_HOST_VARIANT}" "x86_64"
```
Comment (gribozavr) [inline CMakeLists.txt]: Could you replace back this `${SWIFT_HOST_VARIANT}` as well?

File: CMakeLists.txt [hunk 10]
```diff
@@ -435,114 +484,91 @@ function(is_sdk_requested name result_var_name)
...
-
-  # FIXME: This will not work while trying to cross-compile.
-  if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
-    set(SWIFT_HOST_VARIANT_ARCH "x86_64")
-    set(SWIFT_PRIMARY_VARIANT_ARCH_default "x86_64")
-
-    if("${SWIFT_ANDROID_NDK_PATH}" STREQUAL "")
-      set(swift_can_crosscompile_stdlib FALSE)
+  # Calculate the host triple
+  if("${SWIFT_HOST_TRIPLE}" STREQUAL "")
```
Comment (gribozavr) [inline CMakeLists.txt]: Would it be ever set to non-empty?  I'd recommend not supporting this until we have a compelling usecase.

File: CMakeLists.txt [hunk 11]
```diff
@@ -435,114 +484,91 @@ function(is_sdk_requested name result_var_name)
...
+  configure_sdk_unix(CYGWIN "Cygwin" "windows" "cygwin" "${SWIFT_HOST_VARIANT_ARCH}" "${SWIFT_HOST_VARIANT_ARCH}-unknown-windows-cygnus")
 
-  set(SWIFT_HOST_VARIANT_SDK "CYGWIN")
-  set(SWIFT_HOST_VARIANT_ARCH "x86_64")
-  
-  set(SWIFT_PRIMARY_VARIANT_SDK_default "CYGWIN")
-  set(SWIFT_PRIMARY_VARIANT_ARCH_default "x86_64")
-elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
-  # Set defaults.
+elseif("${SWIFT_HOST_VARIANT_SDK}" MATCHES "(OSX|IOS-*|TVOS-*|WATCHOS-*)")
```
Comment (gribozavr) [inline CMakeLists.txt]: Why dashes?  I think valid values are, for example, `WATCHOS` and `WATCHOS_SIMULATOR`, and neither has dashes.

File: CMakeLists.txt [hunk 12]
```diff
@@ -650,9 +679,10 @@ elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
   # set(SWIFT_PRIMARY_VARIANT ${SWIFT_PRIMARY_VARIANT_GUESS} CACHE STRING
   #    "[OSX-DA, OSX-RA, OSX-R, IOS-DA, IOS-RA, IOS-R, IOS_SIMULATOR-DA, IOS_SIMULATOR-RA, IOS_SIMULATOR-R]")
   #
-  # FIXME: hardcode OS X as the default variant for now.
-  set(SWIFT_PRIMARY_VARIANT_SDK_default "OSX")
+  # Override SWIFT_PRIMARY_VARIANT-* on iOS hosts; set to OSX
+  set(SWIFT_PRIMARY_VARIANT_SDK_default  "OSX")
```
Comment (gribozavr) [inline CMakeLists.txt]: Extra whitespace change?

File: CMakeLists.txt [hunk 13]
```diff
@@ -650,9 +679,10 @@ elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
   # set(SWIFT_PRIMARY_VARIANT ${SWIFT_PRIMARY_VARIANT_GUESS} CACHE STRING
   #    "[OSX-DA, OSX-RA, OSX-R, IOS-DA, IOS-RA, IOS-R, IOS_SIMULATOR-DA, IOS_SIMULATOR-RA, IOS_SIMULATOR-R]")
   #
-  # FIXME: hardcode OS X as the default variant for now.
-  set(SWIFT_PRIMARY_VARIANT_SDK_default "OSX")
+  # Override SWIFT_PRIMARY_VARIANT-* on iOS hosts; set to OSX
```
Comment (gribozavr) [inline CMakeLists.txt]: Please end comments with a full stop.

File: CMakeLists.txt [hunk 14]
```diff
@@ -563,6 +590,8 @@ elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
     configure_target_variant(OSX-R  "OS X Release"         OSX R  "Release")
   endif()
 
+  # Compatible cross-compile SDKS for OSX: IOS, IOS_SIMULATOR, TVOS, TVOS_SIMULATOR, WATCHOS, WATCHOS_SIMULATOR (archs hardcoded below).
```
Comment (gribozavr) [inline CMakeLists.txt]: 80 columns.

File: CMakeLists.txt [hunk 15]
```diff
@@ -435,114 +484,91 @@ function(is_sdk_requested name result_var_name)
...
-
-#  set(CMAKE_EXECUTABLE_FORMAT "ELF")
-
+  
+elseif("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "CYGWIN")
+  
+  # set(CMAKE_EXECUTABLE_FORMAT "ELF")
   set(SWIFT_HOST_VARIANT "windows" CACHE STRING
       "Deployment OS for Swift host tools (the compiler) [windows].")
+  configure_sdk_unix(CYGWIN "Cygwin" "windows" "cygwin" "${SWIFT_HOST_VARIANT_ARCH}" "${SWIFT_HOST_VARIANT_ARCH}-unknown-windows-cygnus")
```
Comment (gribozavr) [inline CMakeLists.txt]: 80 columns.

Comment (gribozavr) [inline CMakeLists.txt]: I'd also recommend substituting concrete values for `${SWIFT_HOST_VARIANT_ARCH}` back.

**Comment by gribozavr** [comment]
@swift-ci Please test

File: CMakeLists.txt (hunk 10 above)
Comment (karwa) [inline CMakeLists.txt]: It hasn't been calculated yet. We check if it's empty to support explicitly specifying it (for example, there are the generic `arm` and `armhf` triples - `arm-linux-gnueabi` and  `arm-linux-gnueabi**hf**`). The `SWIFT_HOST_VARIANT_ARCH` is still just `arm`, so we would need the build script, which knows which target we want, to build for a generic `arm` architecture but with the `armhf` triple.

I think that's right - not 100%, but the worst this check does is let you override the host triple.

File: CMakeLists.txt [hunk 16]
```diff
@@ -435,114 +484,92 @@ function(is_sdk_requested name result_var_name)
...
-  # FIXME: This only matches ARMv6l (by far the most common variant).
-  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv6l")
-    configure_sdk_unix(LINUX "Linux" "linux" "linux" "armv6" "armv6-unknown-linux-gnueabihf")
-    set(SWIFT_HOST_VARIANT_ARCH "armv6")
-    set(SWIFT_PRIMARY_VARIANT_ARCH_default "armv6")
-  # FIXME: This only matches ARMv7l (by far the most common variant).
-  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7l")
-    configure_sdk_unix(LINUX "Linux" "linux" "linux" "armv7" "armv7-unknown-linux-gnueabihf")
-    set(SWIFT_HOST_VARIANT_ARCH "armv7")
+    # FIXME: Is this correct? Should Android be the primary variant when it's a cross-compile target?
```
Comment (rintaro) [inline CMakeLists.txt]: If we have a way *not* to configure `SWIFT_HOST_VARIANT_SDK`, it's possible.
But, I think, not for now.
See discusson on #2446 
FWIW, `SWIFT_PRIMARY_VARIANT_{SDK, ARCH}` is not so important for the build.
Typically, it's just used to make alias of build target name.
For instance, "swift-stdlib" is an alias of "swift-stdlib-macosx-x86_64" on Mac OS X

Comment (karwa) [inline CMakeLists.txt]: Okay, I've updated it so that Android is only set as the primary variant if we aren't building for the host. We never actually get configured like that, but that's how we should behave if we were to be.

File: CMakeLists.txt (hunk 4 above)
Comment (karwa) [inline CMakeLists.txt]: Also, to be truly flexible we would need to pass along the sysroot/SDK path for every SDK/arch combination. Darwin targets currently cheat and use `xcrun`, Android passes it in as a specific parameter. That's not flexible - what if I wanted to cross-compile the stdlib for multiple linux targets? Each would have a different sysroot, and we'd need to tell it, for example:

`-DSWIFT_SDKs="LINUX-x86_64;LINUX-armv6;LINUX-armv7;ANDROID-armv7"
-DSWIFT_SDK_LINUX_x86_64_SYSROOT="/"
-DSWIFT_SDK_LINUX_armv6_SYSROOT="/sysroots/debian.jessie.armv6"
-DSWIFT_SDK_LINUX_armv7_SYSROOT="/sysroots/debian.jessie.armv7"
-DSWIFT_SDK_ANDROID_armv7_SYSROOT="/sysroots/android-ndk-armv7"`

**Comment by karwa** [comment]
I believe I've addressed all the feedback now

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
The failures are unrelated.
